### ISSUE TITLE
Planning story: ZigBee prototype

### ISSUE THREAD

**Issue by lmas** [lmas/d0020e_code#2]
- [x] Document the use case, see slide L06.1 pg. 14 (and to the end) for examples
- [x] Draw a diagram to get an overview of what's required to implement?
- [x] ~Split up the use case into stories, document each (in report and as github issues?). See slide L2.1 pg. 11 for an example.~
- [x] Split each story into tasks, create issues for each and estimate work time?
- [x] Prioritise the stories/tasks and put them on the board
- Pick one story and start working on the highest prio. task
  - [x] [Base system](https://github.com/users/lmas/projects/4/views/1?pane=issue&itemId=89034423&issue=lmas%7Cd0020e_code%7C10)
  - [x] [Working API calls](https://github.com/users/lmas/projects/4/views/1?sliceBy%5Bvalue%5D=ZigBeeValve&pane=issue&itemId=89033449&issue=lmas%7Cd0020e_code%7C9)

**Use case `#1` description**
Name: `Continuously update a ZigBee smart thermostat's setting to desired room temperatur`
Participating actors: `Comfortstat, ZigBee device host`
Entry condition: `New setpoint is received from Comfortstat`
Exit condition: `New temperature was sent to smart thermostat`
Event flow: 
- Receive new setpoint from Comfortstat
- If new setpoint differs from old by a small margin:
    - Store new setpoint
    - Send wanted setpoint to the smart thermostat using the API from the device host

**Use case `#2` description**
Name: `Periodically compare room temperature with setpoint and toggle the smart power plug accordingly`
Participating actors: `Comfortstat, ZigBee device host`
Entry condition: `Every x period`
Exit condition: `New state change has been sent to smart plug`
Event flow: 
- Start background process and let it run every x period
- Receive new setpoint from Comfortstat and store it
- For each iteration in the background process:
    - Get current temperature from the ds18b20 system
    - Determine new state of plug, by comparing stored setPoint and temperature
    - If state changed, send new state to smart plug using API from the ZigBee device host

```mermaid
sequenceDiagram
participant SR as ServiceRegistrar
participant OR as Orchestrator
participant CS as Comfortstat
participant TP as Temperature
participant ZB as ZigBee
participant API as ZigBee device host

loop every x seconds
    ZB->>+SR: Register "setpoint" service
    SR-->>-ZB: ok
end

loop New setpoint from Comfortstat
    CS->>+ZB: New setPoint
    ZB-->>-CS: ok

    critical If setpoint is new
        ZB->>+API: Send setpoint to smart thermostat
        API-->>-ZB: ok
    end
end

loop every x period
    ZB->>+OR: Find "temperature" service
    OR->>+SR: Query registry
    SR-->>-OR: System URL
    OR-->>-ZB: Service URL

    ZB->>+TP: Get temperature
    TP-->>-ZB: data

    critical If temperature < setpoint
        ZB->>+API: Send state ON to smart plug
        API-->>-ZB: ok
    option else
        ZB->>+API: Send state OFF to smart plug
        API-->>-ZB: ok
    end
end
```

**Comment by Pake-TU** [comment]
```mermaid
sequenceDiagram
    loop every x seconds
    ZigBee->>ServiceRegistrar: register
    activate ServiceRegistrar
    activate ZigBee
    ServiceRegistrar->>ZigBee: registered
    deactivate ServiceRegistrar
    deactivate ZigBee
    end
    
    loop every x seconds
    Thermostat->>ZigBee: Rotate
    activate ZigBee
    ZigBee->>Thermostat: ok!
    ZigBee->>Radiator: setTemp()
    activate Radiator
    Radiator->>ZigBee: ok!
    deactivate Radiator
    deactivate ZigBee
    end
```

**Comment by vanDeventer** [comment]
This is a great drawing. It lets me reflect on what you are supposed to do.
- I did not give you a hydronic radiator. I gave you a ZigBee valve. I believe it has a thermostat and a temperature sensor in it. All you have to do is to send a set point every hour.
- I did not give you an electric radiator. I have one in my office. I did give you a smart outlet and a temperature sensor. Your thermostat has to find the URL of that sensor and that ZigBee plug. Then you have to turn it on and off. 
- The time series database (InfluxDB as an asset) has to find the address of the signal it wants to log. temperatures, prices, set points.

Let first have a new system Comfor€stat. Its job is to maximize comfort while minimizing costs. It will get the current price of electricity and determine the desired set point for a given room. If the room has a valve, it will tell it that set point. If that room has an electrical heater, it will have to find the eThermostat and PUT its set point. The eThermostat must read the temperature and turn on and off the plug. It can also get the power consumed by the electrical heater, which should be logged.

**Comment by lmas** [comment]
Hmm I think this was a huge case of "poorly choosen names" in the diagram. We're discussing this ATM

**Comment by vanDeventer** [comment]
I do not agree with you. I do not think that you are stupid (at least in this case, which is the only one I have with you). What we are doing now is the most important part of the project: reflect on what we want to build. Only smart people do that. You could have started to build something that would have to be redone. Instead you made a diagram so that we can discuss ideas and views. I am impressed.

**Comment by lmas** [comment]
In between the moment you're sharing, Patrick and Jan, I would like to point out we updated the diagram and added a description of the system after the latest clarification. The diagram needs to be cleaned up with the final names used in the code of course.
There are separate issues with more info for each individual task too.

 Do you have more suggestions or insights?

**Comment by lmas** [comment]
Merged in documentation from #11 and updated description and diagram. Closing this as complete (although not signed off with Jan).

**Comment by vanDeventer** [comment]
I have to look for the diagrams to learn from that.
I have pushed my code to the **dev** branches of https://github.com/vanDeventer/metalepsis and https://github.com/vanDeventer/metonymy 
The big changes that might be interesting for you is the esr system and the ds18b20 system where I make use of channels between the goroutines.
I have to fix the settings in the other GitHub so that I am allowed to push my code there.
